examples/beacon-root/contracts/src/Beacon.sol

```diff
@@ -0,0 +1,54 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.20;
+
+import {Steel} from "risc0/steel/Steel.sol";
+
+/// @title Steel Beacon Chain library
+library Beacon {
+    /// @notice The address of the Beacon roots contract.
+    /// @dev https://eips.ethereum.org/EIPS/eip-4788
+    address internal constant BEACON_ROOTS_ADDRESS = 0x000F3df6D732807Ef1319fB7B8bB8522d0Beac02;
+
+    /// @notice The Beacon block root could not be found as the next block has not been issued yet.
+    error NoParentBeaconBlock();
+
+    /// @notice Attempts to find the root of the Beacon block with the given timestamp.
+    /// @dev Since the Beacon roots contract only returns the parent Beacon block’s root, we need to find the next
+    ///      Beacon block instead. This is done by adding the block time of 12s until a value is returned.
+    function blockRoot(uint256 timestamp) internal view returns (bytes32 root) {
```

Comment:
When would looking up by timestamp be used instead of by hash?

Reply:
I am not sure I understand. The EIP-4788 beacon root contract gives you a way to retrieve the (parent) beacon block root by its respective timestamp. As such, Steel commits to the beacon block root plus the beacon block timestamp, and during on-chain validation we check that the beacon root contract returns the same root for that timestamp.

Reply:
Ok, thank you. That clarifies my question.